lib/Optimizer/Transforms/MultiControlDecomposition.cpp

```diff
@@ -0,0 +1,158 @@
+    ancillas.push_back(builder.create<quake::AllocaOp>(loc));
+
+  SmallVector<Operation *> toCleanup;
+  std::array<Value, 2> cs = {controls[0], controls[1]};
+
+  // Compute intermediate results
+  toCleanup.push_back(builder.create<quake::XOp>(loc, cs, ancillas[0]));
+  if (!negatedControls.empty() &&
+      (negatedControls[0] != false || negatedControls[0] != false))
+    toCleanup.back()->setAttr("negated_qubit_controls",
+                              builder.getDenseBoolArrayAttr(
+                                  {negatedControls[0], negatedControls[1]}));
+  for (size_t c = 2, a = 0, n = controls.size() - numControlsOutput + 1; c < n;
+       ++c, ++a) {
+    cs = {controls[c], ancillas[a]};
+    toCleanup.push_back(builder.create<quake::XOp>(loc, cs, ancillas[a + 1]));
+    if (!negatedControls.empty() && negatedControls[c] != false)
+      toCleanup.back()->setAttr("negated_qubit_controls",
+                                builder.getDenseBoolArrayAttr({true, false}));
+  }
+
+  // Compute output
+  if (numControlsOutput == 1)
+    createOperator(loc, name, parameters, ancillas.back(), targets, builder);
+  else {
+    cs = {controls.back(), ancillas.back()};
+    Operation *out =
+        createOperator(loc, name, parameters, cs, targets, builder);
+    if (!negatedControls.empty() && negatedControls.back() != false)
+      out->setAttr("negated_qubit_controls",
+                   builder.getDenseBoolArrayAttr({true, false}));
+  }
+
+  // Cleanup intermediate results
+  for (Operation *op : llvm::reverse(toCleanup))
+    builder.clone(*op);
+}
+
+//===----------------------------------------------------------------------===//
+// Pass implementation
+//===----------------------------------------------------------------------===//
+
+struct Decomposition
+    : public opt::impl::MultiControlDecompositionPassBase<Decomposition> {
+  using MultiControlDecompositionPassBase::MultiControlDecompositionPassBase;
+
+  void runOnOperation() override {
+    func::FuncOp func = getOperation();
+
+    OpBuilder builder(func);
+    func.walk([&](quake::OperatorInterface op) {
+      // This pass does not handle Quake's value semantics form.
+      if (!quake::isAllReferences(op))
+        return;
+
+      builder.setInsertionPoint(op);
+      SmallVector<Value> controls;
+      SmallVector<bool> negatedControls;
+      if (failed(extractControls(op, controls, negatedControls, builder)))
+        return;
+
+      if (controls.size() == 2) {
+        if (isa<quake::XOp, quake::ZOp>(op))
+          return;
+        Value ancilla = builder.create<quake::AllocaOp>(op->getLoc());
+        auto andOp =
+            builder.create<quake::XOp>(op->getLoc(), controls, ancilla);
+        andOp.setNegatedQubitControls(op.getNegatedControls());
+        createOperator(op->getLoc(), op->getName().getStringRef(),
+                       op.getParameters(), ancilla, op.getTargets(), builder);
+        builder.clone(*andOp);
+        op.erase();
```

Comment:
Why isn't `replaceOpWithNewOp()` being used? That should work for variant forms of the quantum operators (although this rewrite is guarded against that).

Reply:
In this case, I'm using `OpBuilder` instead of the rewriter, and it doesn't have the `replaceOpWithNewOp` method. Also, even if I wanted to use the rewriter, the `replaceOpWithNewOp` method is templated on the operation type. Here I want to avoid this by creating a new operation using only its name (which I can get from the OperatorInterface), so that won't need to change this code when adding new quantum operations to the dialect. (I suppose we could also use some macro shenanigans for the same result, but I would rather not do that.)